Plots in an editor

timtmok opened the pull request:
Address #2270

This adds an action to display a plot from the Plots pane in an editor. It works on static and dynamic plots. It is an editor just for plots and hosts a React container to reuse the components from the Plots pane. So, it could easily display web plots.

A new editor and editor input has been created. The plot id is used for the editor title. It's not very descriptive and perhaps should be replaced with something that makes sense for a user. The plots service now tracks another list of plots for editors. The original plot client is cloned as a new object. It uses the same plot id as the original plot. The service now has an `openEditor()` method that opens the current plot in an editor.

The entire feature is hidden behind an experimental flag. It is in the Application > Experimental category in the preferences. It is off by default until the rest of the editor actions are implemented. See the issue for linked follow-up work.

### QA Notes
It's likely buggy with how the plot behaves with being moved into a new window and splitting the editor. It shares the sizing policy with the Plots pane so changing it there will re-render the editor in the new size.

timtmok committed:
Basic plots editor

timtmok committed:
Display plot in editor

Renders static and dynamic plots in the editor

timtmok committed:
Update open plot editor icon

timtmok committed:
Set editor name to plot id

timtmok committed:
Show plot editor button when applicable

timtmok committed:
Add feature flag

timtmok committed:
Move setting to Application/Experimental

nstrayer reviewed (changes requested):
Looking good and impressive work!

- Left  a few comments on code patterns etc.. 

- I feel like we should consider the phrasing of this feature too, I'm not sure how much the user knows that the tabs are represented by "editors." I left a few comments on the code itself about this but I think just a bit more description will go a long way to alleviating potential confusion. 
 
<img width="680" alt="image" src="https://github.com/user-attachments/assets/aa6912a1-aa5c-4d7d-b05d-63252f0cd1be">
- Probably should have the editor tab show something more meaningful than a random id string.


- I noticed that we are using the grab cursor even though there's no scrolling available. Is this a choice to make it clear there's no right-click action menu stuff or just a left-over from the reused components?

- On the product side a thing I am curious about is if we want to add the same controls that one gets in the plots pane to the editor? E.g. placing the actions toolbar in the editor pane. I think 90% of the usecases will be just looking at the plot larger, but I could imagine people wanting to be able to export etc from there as well. 
This would also have implications on state transfer, e.g. aspect ratio being preserved across transition from the plots pane to the editor tab. 
I'm also good to defer this stuff for a future PR as it could add a lot of work and delay people getting access to this.

timtmok commented on src/vs/workbench/contrib/positronPlots/browser/positronPlotsActions.ts:
```diff
@@ -204,3 +206,31 @@ export class PlotsPopoutAction extends Action2 {
+	constructor() {
+		super({
+			id: PlotsEditorAction.ID,
+			title: localize2('positronPlots.openEditor', 'Open Plot in Editor'),
```
"Editor tab" is better phrasing. It's good to set this right because I hadn't noticed this all this time how misleading it can be!

timtmok commented on src/vs/workbench/contrib/positronPlotsEditor/browser/editorPlotsContainer.tsx:
```diff
@@ -0,0 +1,52 @@
+}
+
+export const EditorPlotsContainer = (props: EditorPlotsContainerProps) => {
+	usePositronPlotsContext();
```
I'll fix that. I forgot to remove it when I refactored how to display the plot.

timtmok commented on src/vs/workbench/contrib/positronPlotsEditor/browser/editorPlotsContainer.tsx:
```diff
@@ -0,0 +1,52 @@
+
+export const EditorPlotsContainer = (props: EditorPlotsContainerProps) => {
+	usePositronPlotsContext();
+	const render = (plotClient?: IPositronPlotClient) => {
```
I'll fix it!

timtmok commented on src/vs/workbench/contrib/positronPlotsEditor/browser/positronPlotsEditor.contribution.ts:
```diff
@@ -0,0 +1,97 @@
+			`${Schemas.positronPlotsEditor}:**/**`,
+			{
+				id: PositronPlotsEditorInput.EditorID,
+				label: localize('positronPlotsEditor', 'Plots Editor'),
```
Good to clarify the name. It's not that much longer so it should look fine wherever it might show up in the UI.

timtmok commented on src/vs/workbench/contrib/positronPlotsEditor/browser/positronPlotsEditor.tsx:
```diff
@@ -0,0 +1,215 @@
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+		@IContextKeyService private readonly _contextKeyService: IContextKeyService,
+		@IContextMenuService private readonly _contextMenuService: IContextMenuService,
+		@IStorageService storageService: IStorageService,
+		@ITelemetryService telemetryService: ITelemetryService,
+		@IThemeService themeService: IThemeService,
```
Good point! It makes sense to remove `_` in the editor contribution.

timtmok committed:
Apply suggestions from code review

Co-authored-by: Nick Strayer <nick.strayer@posit.co>
Signed-off-by: Tim Mok <timtmok@users.noreply.github.com>

timtmok committed:
Review changes

jmcphers reviewed (changes requested):
<img width="996" alt="image" src="https://github.com/user-attachments/assets/c6ebb368-00df-4e92-9eb7-e494d51eeda4">

Because this approach uses only one comm, we need to avoid disposing/closing the comm until all instances of the plot in the UI have been closed.

As it stands, cleaning up the plot from the Plots pane causes it to stop working in the editor tab; attempting to resize the editor tab results in a progress bar drawn right in the middle of the plot that never dismisses since the backend can't render the plot any more. 

I also observed a fair bit of cross-talk between the plot in the editor and the plot in the Plots pane -- sometimes rendering it in one place caused both places to show a progress bar and/or the resized image to show up in both places. Might be expected at this stage?

Another issue is that you lose your editor plots on reload. You don't need to solve this right away, but we should have a plan to do it. See #3172 for context.

timtmok commented:
Pushing a rebase on `main`.

It looks like it's best to create a new comm backed by the same plot in the backend. It will require updating Ark as well but this should simplify the frontend. It turns out sharing a comm on the frontend makes it far more complicated with render queues, throttling, and managing the render events.

jmcphers commented:
If we move this complexity to the backend we will also need to implement for Python and all future language packs. I'd rather have the front end handle all of the complexity around events/throttling since it will keep the logic in one place and make things simpler at the API level.

I'd imagine this could work by adding a service that can manage the plot comms, adding and removing UI instances of the plot (e.g. for the editor placement, the Plots pane, the Save dialog, etc.) and maintaining a single render queue but only delivering events/RPC results to the UI instance that owned the request. When the UI instance refcount drops to zero, we can dispose the comm. 

Willing to be convinced otherwise, lmk if you want to discuss realtime!

timtmok committed:
Create plot client in backend

Opens another comm to a plot in the backend
Implement Python plots in editor for the frontend

timtmok committed:
Revert "Create plot client in backend"

This reverts commit 5da4fa88c85675ff22b5f4918580432c523ac1f4.

timtmok committed:
Shared plot comm

timtmok committed:
Initial comm cleanup when removing plots

timtmok committed:
Handle comm close on editor close

timtmok committed:
Fix unregisterPlotClient

timtmok committed:
Clean up preview calls

Fix rebase from main

timtmok committed:
Fix disposing editor's plot client

timtmok commented:
This is now at a point where the `PlotClientInstance` has been refactored with the calls to the comm pulled out into `PositronPlotCommProxy`. This allows sharing the comm while the plot clients can request renders with their own parameters.

The plots service tracks how many clients are still using the comm and will dispose the comm when no more clients are using it. It still uses a new plot client list for editors. I do think it's worth following up with another change to go back to one list by updating `IPositronPlotMetadata` with a location. This should make it easy to save the plot state for editors and restore it when the UI reloads.

timtmok committed:
Fix plot client disposed too early

Use editor input dispose to determine the editor is finished with the plot client

jmcphers commented on src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts:
```diff
@@ -209,6 +118,9 @@ export class PlotClientInstance extends Disposable implements IPositronPlotClien
 	private readonly _renderUpdateEmitter = new Emitter<IRenderedPlot>();
 
+	/**
+ * Event that fires when the plot wants to display itself.
```
Nit: indentation

jmcphers commented on src/vs/workbench/services/languageRuntime/common/positronPlotCommProxy.ts:
```diff
@@ -0,0 +1,267 @@
+
+export class PositronPlotCommProxy extends Disposable {
+	/**
+ * The currently active render request, if any.
```
Nit: indentation

jmcphers commented on src/vs/workbench/services/languageRuntime/common/positronPlotCommProxy.ts:
```diff
@@ -0,0 +1,267 @@
+	private _currentRender?: DeferredRender;
+
+	/**
+ * The underlying comm
```
Nit: indentation

jmcphers reviewed (dismissed):
Looks good! I left some notes re: indentation but there are a bunch more places where the indentation got garbled (maybe an editor setting?).

I was able to verify that the newest set of changes address the lifecycle management issues; i.e. removing a plot from the Plots pane doesn't cause it to break in the Editor, and closing the Editor cleans up the comm correctly if it is the last remaining reference.

There are still some drawing issues; the progress bar doesn't seem to draw consistently in the editor pane, and when it does, it draws behind the plot. 

<img width="856" alt="image" src="https://github.com/user-attachments/assets/914c23c0-be2d-49ff-a648-68e30ee93bb7">

There's also some crosstalk w/r/t the sizing policy; if you change the sizing policy in the Plots pane, it redraws the plot in the Editor pane too, which feels a little jarring. No need to try to address that now; we will eventually just want the editor plot tab to have its own toolbar where these things can be controlled/surfaced.

timtmok commented:
I noticed the progress misplacement as well. I opened https://github.com/posit-dev/positron/issues/4562 to address it later.

timtmok committed:
Formatting

timtmok committed:
Reconnect comm events

timtmok commented:
I noticed that the comm events weren't hooked up in the comm proxy. It was preventing the plot client from disposing when the comm closed.

The editor won't close when the comm closes. I think a follow up can address this. It can do the same thing as the data explorer and show a message that the plot is gone.

nstrayer commented on src/vs/workbench/contrib/positronPlotsEditor/browser/positronPlotsEditor.contribution.ts:
```diff
@@ -0,0 +1,97 @@
+		[POSITRON_EDITOR_PLOTS]: {
+			scope: ConfigurationScope.APPLICATION,
+			type: 'boolean',
+			default: true,
```
Should we switch this to `false`?

timtmok replied:
Good catch. I had assumed this was what I set it to!

nstrayer reviewed (changes requested):
This is awesome and I'm excited to have this! 

I like the new shape of things and everything works as expected. 

Just had a few nits and general comments about react code organization/patterns that can be addressed now or later (or ignored if they're off base!)

Main stopper is the feature flag default, but should be good to go after that!

timtmok committed:
Apply suggestions from code review

Co-authored-by: Nick Strayer <nick.strayer@posit.co>
Signed-off-by: Tim Mok <timtmok@users.noreply.github.com>

timtmok committed:
Review changes

timtmok commented on src/vs/workbench/contrib/positronPlots/browser/components/actionBars.tsx:
```diff
@@ -159,6 +173,18 @@ export const ActionBars = (props: PropsWithChildren<ActionBarsProps>) => {
 								ariaLabel={localize('positron-popout-plot', "Open plot in new window")}
 								onPressed={popoutPlotHandler} />
 						}
+						{enableEditorPlot &&
```
This is a good pattern to remember for conditional rendering! I've fixed up the other instances as well.

timtmok commented on src/vs/workbench/contrib/positronPlots/browser/positronPlotsService.ts:
```diff
@@ -1001,6 +1016,101 @@ export class PositronPlotsService extends Disposable implements IPositronPlotsSe
+		const commProxy = new PositronPlotCommProxy(client, metadata);
+		this._plotCommProxies.set(metadata.id, commProxy);
+
+		commProxy.onDidClose(() => {
```
Yes, I've gone through the other places that need to register the disposable

timtmok commented on src/vs/workbench/contrib/positronPlots/browser/positronPlotsService.ts:
```diff
@@ -1001,6 +1016,101 @@ export class PositronPlotsService extends Disposable implements IPositronPlotsSe
+		let plotClients = this._plotClientsByComm.get(comm.metadata.id);
+
+		if (!plotClients) {
+			plotClients = new Array<PlotClientInstance>();
```
No reason but I can see now that it can be inferred when creating `plotClients`

timtmok commented on src/vs/workbench/contrib/positronPlotsEditor/browser/editorPlotsContainer.tsx:
```diff
@@ -0,0 +1,50 @@
+}
+
+export const EditorPlotsContainer = (props: EditorPlotsContainerProps) => {
+	const render = (plotClient: IPositronPlotClient) => {
```
Makes sense and I must have overlooked it when initially writing this

timtmok committed:
Fix rendering plot element

timtmok commented on src/vs/workbench/contrib/positronPlotsEditor/browser/positronPlotsEditor.tsx:
```diff
@@ -0,0 +1,219 @@
+		}
+
+		return null;
+	};
```
Just an omission when I created `EditorPlotsContainer` and should be removed.

timtmok commented on src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts:
```diff
@@ -209,6 +118,9 @@ export class PlotClientInstance extends Disposable implements IPositronPlotClien
 	private readonly _renderUpdateEmitter = new Emitter<IRenderedPlot>();
 
+	/**
+	 * Event that fires when the plot wants to display itself.
```
I've reworded it. The plot doesn't want to display itself, it's the runtime that sent an event to show the plot. I think we only care that the plots view shows or switches to showing this plot. We could also make the editor tab associated with this plot the active editor but I think it would be good to see if users want that.

timtmok committed:
Refactor need for useCallback hook

timtmok commented on src/vs/workbench/contrib/positronPlots/browser/components/actionBars.tsx:
```diff
@@ -85,9 +87,21 @@ export const ActionBars = (props: PropsWithChildren<ActionBarsProps>) => {
+				setEnableEditorPlots(isEditorPlotsEnabled());
+			}
+		});
+		setEnableEditorPlots(isEditorPlotsEnabled());
```
Removed the useCallback hook and kept the state to tracking the configuration value. I think this does look cleaner.

nstrayer commented on src/vs/workbench/contrib/positronPlots/browser/components/actionBars.tsx:
```diff
@@ -150,30 +150,32 @@ export const ActionBars = (props: PropsWithChildren<ActionBarsProps>) => {
-							onPressed={copyPlotHandler} />}
-						{enableZoomPlot && <ZoomPlotMenuButton actionHandler={zoomPlotHandler} zoomLevel={props.zoomLevel} />}
-						{enableSizingPolicy &&
+						{(enableSizingPolicy || enableSavingPlots || enableZoomPlot || enablePopoutPlot) ? <ActionBarSeparator /> : null}
```
super-nit: I would maybe extract the long `or` logic here to a well-named variable to make it clearer why something is being rendered or not.

nstrayer commented on src/vs/workbench/contrib/positronPlots/browser/components/actionBars.tsx:
```diff
@@ -59,7 +59,7 @@ export interface ActionBarsProps {
 	// Hooks.
 	const positronPlotsContext = usePositronPlotsContext();
-	const [enableEditorPlot, setEnableEditorPlots] = React.useState<boolean>(false);
+	const [useEditorPlots, setUseEditorPlots] = React.useState<boolean>(positronPlotsEditorEnabled(props.configurationService));
```
Good call on the initial value to avoid a potentially unnecessary render.

nstrayer reviewed (approved):
Looking good -- this has shaped up real nicely!
